web-common/src/features/dashboards/state-managers/loaders/most-recent-explore-state.ts

```diff
@@ -0,0 +1,159 @@
+  "allMeasuresVisible",
+  "visibleMeasureKeys",
+  "allDimensionsVisible",
+  "visibleDimensionKeys",
+  "leaderboardMeasureName",
+  "sortDirection",
+  "leaderboardContextColumn",
+];
+// Keys that do not need any special handling and can be directly copied over
+const DirectCopyExploreStateKeys: (keyof MetricsExplorerEntity)[] = [
+  "showTimeComparison",
+  ...ExploreViewKeys,
+];
+// Keys that are not defined in certain views
+const ExploreStateNonDefinedInViewKeys: Record<
+  DashboardState_ActivePage,
+  (keyof MetricsExplorerEntity)[]
+> = {
+  [DashboardState_ActivePage.UNSPECIFIED]: [],
+  [DashboardState_ActivePage.DEFAULT]: [],
+  [DashboardState_ActivePage.DIMENSION_TABLE]: [],
+  [DashboardState_ActivePage.TIME_DIMENSIONAL_DETAIL]: ExploreViewKeys,
+  [DashboardState_ActivePage.PIVOT]: ExploreViewKeys,
+};
+
+function getKeyForLocalStore(
+  exploreName: string,
+  storageNamespacePrefix: string | undefined,
+) {
+  return `rill:app:explore:${storageNamespacePrefix ?? ""}${exploreName}`.toLowerCase();
+}
+
+export function getMostRecentExploreState(
+  exploreName: string,
+  storageNamespacePrefix: string | undefined,
+  metricsView: V1MetricsViewSpec,
+  explore: V1ExploreSpec,
+) {
+  try {
+    const rawUrlSearch = localStorage.getItem(
+      getKeyForLocalStore(exploreName, storageNamespacePrefix),
+    );
+    if (!rawUrlSearch) return { partialExploreState: undefined, errors: [] };
+
+    return convertURLSearchParamsToExploreState(
+      new URLSearchParams(rawUrlSearch),
+      metricsView,
+      explore,
+      // Send empty preset so that fields are always stored.
+      {},
+    );
+  } catch {
+    // no-op
+  }
+  return { partialExploreState: undefined, errors: [] };
+}
+
+export function saveMostRecentExploreState(
+  exploreName: string,
+  storageNamespacePrefix: string | undefined,
+  metricsView: V1MetricsViewSpec,
+  explore: V1ExploreSpec,
+  timeControlsState: TimeControlState | undefined,
+  exploreState: MetricsExplorerEntity,
+) {
+  const { partialExploreState: existingExploreState } =
+    getMostRecentExploreState(
+      exploreName,
+      storageNamespacePrefix,
+      metricsView,
+      explore,
+    ) ?? {};
+  const newExploreState: Partial<MetricsExplorerEntity> = {};
+
+  DirectCopyExploreStateKeys.forEach((k) => {
+    (newExploreState as any)[k] = exploreState[k];
+  });
+  if (existingExploreState) {
+    ExploreStateNonDefinedInViewKeys[exploreState.activePage].forEach((k) => {
+      (newExploreState as any)[k] = existingExploreState[k];
+    });
+  }
+  newExploreState.activePage = DashboardState_ActivePage.DEFAULT;
+
+  // Since we are storing a few settings in timeControlsState, url params is populated using it.
+  // Hopefully we will store everything in a single place in the future and we can update timeControlsState directly.
+  if (timeControlsState) {
+    // Custom handling for time range. We are retaining the previous range if the current range is a custom range.
```

Comment:
Is this a product requirement (CC @mindspank)? This file would be much simpler if we were just storing the whole explore state as-is